dap_job_quality/utils/create_df_for_salary_analysis.py

```diff
@@ -0,0 +1,103 @@
+from dap_job_quality.getters.data_getters import load_s3_data, load_s3_excel, save_to_s3
+import pandas as pd
+
+
+# Current OJO data paths
+occupation_measures_s3 = "outputs/data/ojo_application/extracted_green_measures/20240213/ojo_large_sample_occupation_green_measures_production_true.csv"
+large_ojo_sample_s3 = (
+    "outputs/data/ojo_application/deduplicated_sample/all_ojo_sample.parquet"
+)
+location_measures_s3 = (
+    "outputs/data/ojo_application/deduplicated_sample/all_locations_data_sample.parquet"
+)
+salary_measures_s3 = (
+    "outputs/data/ojo_application/deduplicated_sample/all_salaries_data_sample.parquet"
+)
+
+
+def merge_ojo_df(
+    ojo_df: pd.DataFrame,
+    ojo_occ: pd.DataFrame,
+    ojo_loc: pd.DataFrame,
+    ojo_sal: pd.DataFrame,
+    save_file=False,
+) -> pd.DataFrame:
+    """Takes de-duplicated OJO data and returns a dataframe with the relevant data
+    (ITL1 codes, SOC2 and SOC4 codes, salary min, salary max, salary mean, and job title)
+    """
+
+    # Merge dataframes
+    ojo_df = ojo_df.merge(
+        ojo_loc[["id", "itl_1_code", "itl_1_name"]], on="id", how="left"
+    )  # Get ITL1 code and name
+    ojo_df = ojo_df.merge(
+        ojo_occ[["job_id", "SOC"]], left_on="id", right_on="job_id", how="left"
+    ).drop(
+        columns="job_id"
+    )  # Get raw SOC level 4 mapping
+    ojo_df = ojo_df.merge(
+        ojo_sal, on="id", how="left"
+    )  # Min and max annualised salaries
+
+    # Get mean salaries
+    ojo_df["mean_salary"] = ojo_df[
+        ["min_annualised_salary", "max_annualised_salary"]
+    ].mean(axis=1)
+
+    # Get SOC 4 digit codes in single column
+    ojo_df["SOC_clean"] = ojo_df["SOC"].fillna("None")
+    ojo_df["SOC_clean"] = ojo_df["SOC_clean"].apply(lambda x: x.replace("nan", "None"))
+    ojo_df["SOC_clean"] = ojo_df["SOC_clean"].apply(
+        lambda x: eval(x) if isinstance(x, str) else x
+    )
+    ojo_df["SOC_4digit"] = ojo_df["SOC_clean"].apply(
+        lambda x: x.get("SOC_2020") if isinstance(x, dict) else x
+    )
+    ojo_df["SOC_2digit"] = ojo_df["SOC_4digit"].apply(
+        lambda x: x[0:2] if isinstance(x, str) else x
+    )
+
+    # Check all values have carried over from the SOC extraction
+    soc_columns = [col for col in ojo_df.columns if "SOC" in col]
+
+    for col in soc_columns:
+        isnull_counts = ojo_df[soc_columns].isnull().sum()
+        all_counts_same = isnull_counts.nunique() == 1
+
+    if all_counts_same:
+        logger.info("SOC successfully extracted - no additional missing values")
+    else:
+        logger.info("Some additional missing values - check if SOC extracted correctly")
+
+    if save_file == True:
+        save_to_s3(
+            BUCKET_NAME, ojo_df, "job_quality/salary_analysis/salary_analysis_df.csv"
+        )
+        logger.info(
+            f"File saved to {BUCKET_NAME}/job_quality/salary_analysis/salary_analysis_df.csv"
+        )
+
+    return ojo_df
```

Comment:
This df has NAs for the salary in about 25% of rows. The file `outputs/data/ojo_application/deduplicated_sample/all_salaries_data_sample.parquet` is 3,382,859 rows while the cleaned salaries df has information for 4,653,782 unique job adverts. Did you know about this / does Jack know? Does this mean we're missing salary information for some job adverts (about 1 million)?